src/transformStateWithClones.js

```diff
@@ -8,6 +8,36 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  const transformedState = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let nextState = { ...currentState };
+
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(nextState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete nextState[key];
+        }
```

Comment:
Move this code to separated function outside main function `transformStateWithClones`.